projects/hathor-wallet-headless/0005-early-send-tx-lock-release.md

```diff
@@ -38,37 +38,42 @@ The methods that send transactions are:
 - `createNanoContractTransaction`
 - `createAndSendNanoContractTransaction`
 
-These methods (aside from `handleSendPreparedTransaction`) receive an `options` argument, we will add an `autorun` boolean that defaults to `true`.
-When `autorun` is `false` the method should return the `SendTransaction` instance instead.
+We will create new methods that return the `SendTransaction` instance instead and these methods will be changed to use the new methods to avoid code duplication.
 This will give the caller more control over the execution of the "send transaction" process.
 
 ## Headless send-tx lock
 
-We will change the call of any methods listed above to not use autorun.
+We will change the call of any methods listed above to use the new method.
 This will enable the headless to unlock the send-tx lock before mining the transaction.
-Any UTXOs locked during the `prepare-tx` step will not be chosen if another method tries to send a transaction since the `SendTransaction` facade marks them as 'selected as input'.
+
+The send-tx process can be seen as divided in 3 steps:
+
+- `prepare-tx`
+  - Choose UTXOs and create a `Transaction` instance
+  - `SendTransaction` can be instantiated with the `Transaction` instance
+- `mine-tx`
+  - Lock utxos from the `Transaction` instance
+  - Send transaction to a tx-mining service and get the `nonce`
+- `push-tx`
+  - Send the transaction to the fullnode.
+
+The headless will have to manually lock the UTXOs after the `prepare-tx` step so the UTXOs are not chosen by another call to create a transaction.
```

Comment:
Why? Why don't we lock the UTXO's in the `prepare-tx` step?

Reply:
Some facade methods (e.g. `mintTokens`) creates a prepared `Transaction` instance and then call `runFromMining`, which skips the prepare-tx step since the tx is already prepared.

The wallet-lib currently calls the `updateOutputSelected` in the `mineTx` so we lock utxos chosen in the `SendTransaction` and inputs from a prepared `Transaction` instance.

The headless will have to manually call the `updateOutputSelected` method before releasing the send-tx lock to ensure that these utxos are not selected by the next call.